# Pendahuluan
- Virus Corona merubah pola hidup semua orang, dimana mereka lebih memilih untuk menghabiskan waktunya di rumah untuk membaca buku. Sebagai Data Analyst pada sebuah perusahaan start up yang membuat aplikasi pecinta buku, persaingan dalam industri ini, mengharuskan perusahaan untuk membuat analysis berdasarkan database yang berisi buku, penerbit, penulis, serta rating dan ulasan pelanggan atas buku terkait. Analysis yang di dapat bisa untuk memberikan penawaran harga bagi produk baru.

# Tujuan

1. Mengetahui jumlah buku yang di rilis semenjak 1 january 2000
2. Mengetahui jumlah ulasan pengguna dan rating rata-rata untuk setiap buku.
3. Menentukan  penerbit yang telah menerbitkan buku terbanyak dengan jumlah lebih dari 50 halaman (hal ini akan membantu kita dalam mengecualikan brosur dan publikasi serupa dari analisis Anda).
4.  penerbit yang telah menerbitkan buku terbanyak dengan jumlah lebih dari 50 halaman (hal ini akan membantu kita dalam mengecualikan brosur dan publikasi serupa dari analisis Anda).
5. Hitung jumlah rata-rata teks ulasan di antara pengguna yang memberi rating lebih dari 50 buku.

# Tahapan

- Ihtisar data
- Pra proses data
- Analaysis data


# Tugas 

- Hitung jumlah buku yang dirilis setelah tanggal 1 Januari 2000.
- Hitung jumlah ulasan pengguna dan rating rata-rata untuk setiap buku.
- Tentukan penerbit yang telah menerbitkan buku terbanyak dengan jumlah lebih dari 50 halaman (hal ini akan membantu Anda dalam mengecualikan brosur dan publikasi serupa dari analisis Anda).
- Tentukan penulis dengan rating rata-rata buku tertinggi: temukan buku dengan rating minimal 50.
- Hitung jumlah rata-rata teks ulasan di antara pengguna yang memberi rating lebih dari 50 buku.

# Instruksi menyelesaikan tugas

- Jelaskan tujuan dari studi yang Anda lakukan.
- Pelajari tabelnya (tampilkan baris-baris pertamanya).
- Buatlah kueri SQL untuk setiap tugas Anda.
- Jelaskan kesimpulan Anda untuk setiap tugas.
- Buatlah kueri SQL untuk setiap tugas Anda.

# Ihktisar Data

In [32]:
import pandas as pd
from sqlalchemy import create_engine

db_config = {'user': 'practicum_student',         # nama pengguna
             'pwd': 's65BlTKV3faNIGhmvJVzOqhs', # kata sandi
             'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
             'port': 6432,              # port koneksi
             'db': 'data-analyst-final-project-db'}          # nama database

connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
                                                                     db_config['pwd'],
                                                                       db_config['host'],
                                                                       db_config['port'],
                                                                       db_config['db'])

engine = create_engine(connection_string, connect_args={'sslmode':'require'}) 

# Melihat Persebaran Data yang ada di table 

In [33]:
a('''SELECT *
FROM books''',
 con = engine)

book_id  author_id                                              title  \
0          1        546                                       'Salem's Lot   
1          2        465                 1 000 Places to See Before You Die   
2          3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3          4         82  1491: New Revelations of the Americas Before C...   
4          5        125                                               1776   
..       ...        ...                                                ...   
995      996        571           Wyrd Sisters (Discworld  #6; Witches #2)   
996      997        454                        Xenocide (Ender's Saga  #3)   
997      998        201                                    Year of Wonders   
998      999         94                        You Suck (A Love Story  #2)   
999     1000        509  Zen and the Art of Motorcycle Maintenance: An ...   

     num_pages publication_date  publisher_id  
0          594       2005-11-01            93  
1          992       2003-05-22           336  
2          322       2010-12-21           135  
3          541       2006-10-10           309  
4          386       2006-07-04           268  
..         ...              ...           ...  
995        265       2001-02-06           147  
996        592       1996-07-15           297  
997        358       2002-04-30           212  
998        328       2007-01-16           331  
999        540       2006-04-25           143  

[1000 rows x 6 columns]

In [34]:
a('''SELECT *
FROM authors''',
 con = engine)

author_id                                author
0            1                            A.S. Byatt
1            2        Aesop/Laura Harris/Laura Gibbs
2            3                       Agatha Christie
3            4                         Alan Brennert
4            5              Alan Moore/David   Lloyd
..         ...                                   ...
631        632         William Strunk Jr./E.B. White
632        633                           Zadie Smith
633        634                 Zilpha Keatley Snyder
634        635                    Zora Neale Hurston
635        636  Åsne Seierstad/Ingrid Christopherson

[636 rows x 2 columns]

In [35]:
a('''SELECT *
FROM publishers''',
 con = engine)

publisher_id                          publisher
0               1                                Ace
1               2                           Ace Book
2               3                          Ace Books
3               4                      Ace Hardcover
4               5  Addison Wesley Publishing Company
..            ...                                ...
335           336         Workman Publishing Company
336           337                         Wyatt Book
337           338              Yale University Press
338           339                           Yearling
339           340                     Yearling Books

[340 rows x 2 columns]

In [36]:
a('''SELECT *
FROM ratings''',
 con = engine)

rating_id  book_id        username  rating
0             1        1      ryanfranco       4
1             2        1   grantpatricia       2
2             3        1    brandtandrea       5
3             4        2        lorichen       3
4             5        2     mariokeller       2
...         ...      ...             ...     ...
6451       6452     1000  carolrodriguez       4
6452       6453     1000         wendy18       4
6453       6454     1000      jarvispaul       5
6454       6455     1000           zross       2
6455       6456     1000         fharris       5

[6456 rows x 4 columns]

In [37]:
a('''SELECT *
FROM reviews''',
 con = engine)

review_id  book_id        username  \
0             1        1    brandtandrea   
1             2        1      ryanfranco   
2             3        2        lorichen   
3             4        3   johnsonamanda   
4             5        3     scotttamara   
...         ...      ...             ...   
2788       2789      999      martinadam   
2789       2790     1000         wknight   
2790       2791     1000  carolrodriguez   
2791       2792     1000         wendy18   
2792       2793     1000      jarvispaul   

                                                   text  
0     Mention society tell send professor analysis. ...  
1     Foot glass pretty audience hit themselves. Amo...  
2     Listen treat keep worry. Miss husband tax but ...  
3     Finally month interesting blue could nature cu...  
4     Nation purpose heavy give wait song will. List...  
...                                                 ...  
2788  Later hospital turn easy community. Fact same ...  
2789   Change lose answer close pressure. Spend so now.  
2790  Authority go who television entire hair guy po...  
2791  Or western offer wonder ask. More hear phone f...  
2792  Republican staff bit eat material measure plan...  

[2793 rows x 4 columns]

## Melakukan perhitungan release buku setelah tanggal 1 january 2000

In [38]:
a('''
SELECT COUNT(title)
FROM books
WHERE publication_date > '2000-01-01' ''',
con = engine) 

count
0    819

### Kesimpulan awal 1
- Terdapat 819 buku yang release setelah tanggal 1 January 2000

## Melakukan perhitungan jumlah ulasan dan rating rata rata untuk setiap buku 

In [39]:
a('''SELECT reviews.book_id, count(reviews.review_id), avg(ratings.rating) 
FROM reviews
LEFT JOIN ratings on reviews.book_id = ratings.book_id
GROUP BY reviews.book_id  ''',
 con = engine)

book_id  count       avg
0        652      4  4.500000
1        273      4  4.500000
2         51     60  4.250000
3        951      4  4.000000
4        839     28  4.285714
..       ...    ...       ...
989       64     52  4.230769
990       55      4  5.000000
991      148     21  3.428571
992      790      4  3.500000
993      828      4  3.000000

[994 rows x 3 columns]

### Kesimpulan awal 2
- Jika di lihat secara sekilas dan terdapat buku dengan book_id 51 pernah di review selama 60 kali dengan rating rata rata 4.25
- book_id 64 di review sebanyak 60 kali dengan rata rata rating 4.23

## Tentukan penerbit yang telah menerbitkan buku terbanyak dengan jumlah lebih dari 50 halaman

In [40]:
a(''' SELECT publisher, count(publisher) as publ
FROM publishers
LEFT JOIN  books on publishers.publisher_id = books.publisher_id
WHERE books.num_pages > 50
GROUP BY publishers.publisher
ORDER BY publ desc''', 
 con = engine)

publisher  publ
0                                        Penguin Books    42
1                                              Vintage    31
2                             Grand Central Publishing    25
3                                     Penguin Classics    24
4                                     Ballantine Books    19
..                                                 ...   ...
329                                         Turtleback     1
330  Atheneum Books for Young Readers: Richard Jack...     1
331                                     Penguin Signet     1
332                                    Victor Gollancz     1
333                      Harvard Business Review Press     1

[334 rows x 2 columns]

### Kesimpulan awal 3
- Publisher bernama Penguin Books memiliki 42 buku yang memiliki 50 halaman lebih 
- Disusul dengan publisher bernama Vintage yang memiliki 31 buku dengan 50 halaman lebih


## Tentukan penulis dengan rating rata-rata buku tertinggi: temukan buku dengan rating minimal 50.

In [41]:
a('''SELECT authors.author, avg(ratings.rating) as avg_rating, count(ratings.rating) as count_rating
FROM books 
LEFT JOIN authors on books.author_id = authors.author_id
LEFT JOIN ratings on books.book_id = ratings.book_id
GROUP BY authors.author
HAVING count(ratings.rating) >= 50
ORDER BY count_rating desc, avg_rating desc
''',
  con = engine)

author  avg_rating  \
0                          J.K. Rowling/Mary GrandPré    4.288462   
1                                      J.R.R. Tolkien    4.240964   
2                                     Stephenie Meyer    3.662500   
3                                           Dan Brown    3.741259   
4                                     Nicholas Sparks    3.882883   
5                                        Stephen King    4.009434   
6                                       J.D. Salinger    3.846939   
7                                        Rick Riordan    4.130952   
8                                        Jodi Picoult    3.881579   
9             George Orwell/Boris Grabnar/Peter Škerl    3.729730   
10                                     John Steinbeck    3.643836   
11                                    William Golding    3.901408   
12                                       John Grisham    3.971429   
13  William Shakespeare/Paul Werstine/Barbara A. M...    3.787879   
14                                         Lois Lowry    3.738462   
15                                    James Patterson    3.859375   
16                           Roald Dahl/Quentin Blake    4.209677   
17                                    Sophie Kinsella    3.877193   
18           Paulo Coelho/Alan R. Clarke/Özdemir İnce    3.789474   
19                                      Arthur Golden    4.107143   
20                                  Louisa May Alcott    4.203704   
21                                    Agatha Christie    4.283019   
22                  Markus Zusak/Cao Xuân Việt Khương    4.264151   
23                                     Diana Gabaldon    4.300000   

    count_rating  
0            312  
1            166  
2            160  
3            143  
4            111  
5            106  
6             98  
7             84  
8             76  
9             74  
10            73  
11            71  
12            70  
13            66  
14            65  
15            64  
16            62  
17            57  
18            57  
19            56  
20            54  
21            53  
22            53  
23            50

## Hitung jumlah rata-rata teks ulasan di antara pengguna yang memberi rating lebih dari 50 buku.

In [42]:
a('''SELECT count(rating), username
FROM ratings
GROUP BY username 
HAVING count(rating) > 50''',
 con = engine)

count        username
0     55     sfitzgerald
1     53  jennifermiller
2     51          xdavis
3     56          paul88
4     56      martinadam
5     55       richard89

#### Notes 
- Untuk menemukan jumlah rata rata teks review di antara pengguna
- Pertama Membuat tabel yang berisi pengguna yang sudah memberikan rating kepada buku yang lebih dari 50 kali 

In [43]:
a('''SELECT username, count(text) as txt_count
FROM reviews
GROUP by username 
ORDER BY txt_count desc
''',
 con = engine)

username  txt_count
0           susan85         29
1       sfitzgerald         28
2        martinadam         27
3       lesliegibbs         26
4            eallen         26
..              ...        ...
155          todd65         10
156        joseph14         10
157        dwayne09         10
158        robert00          9
159  millernicholas          5

[160 rows x 2 columns]

#### Notes
- Membuat tabel ke dua yang berisi username dan text yang di count, sebagai paramater rata rata review yang di buat oleh pengguna.
- Contoh susan85 sudah melakukan review sebanyak 29 kali

In [44]:
a('''WITH rate_50 as (
    SELECT count(rating) as rating_count, username
    FROM ratings
    GROUP BY username 
    HAVING count(rating) > 50
)
, main as (
    SELECT reviews.username, count(reviews.text) as review_count
    FROM reviews
    INNER JOIN rate_50 on reviews.username = rate_50.username
    GROUP BY reviews.username 
    
)
select ROUND (avg(review_count)) as avg_txt_review
from main''',
  con = engine)

avg_txt_review
0            24.0

#### Notes
- Penggabungan table yang di namakan CTE (Common Table Expression)
- Sebagai contoh melakukan table yang sudah di declare seperti yang di lakukan bernama rate_50

#### Kesimpulan awal 
- Buku yang sudah mendapatkan rating lebih dari 50 kali mempunyai jumlah rata rata review(txt_review) sebanyak 24 kali, atau 24 review oleh pengguna.

# Kesimpulan

- Buku yang di keluarkan setelah 1 january 2020 adalah 819 buku
- Melakukan melakukan perhitungan jumlah ulasan dan rating rata rata untuk setiap buku dapat di lihat terdapat buku dengan book_id 51 pernah di review selama 60 kali dengan rating rata rata 4.25
- Publisher bernama Pinguin Books me release buku yang paling banyak dengan jumlah 42 buku yang merupakan buku lebih dari 50 halaman.
- penulis dengan rating rata-rata buku tertinggi dimana memiliki buku dengan rating minimal 50 adalah JK Rowling di susul oleh JRR TOLKIEN
- Buku yang sudah mendapatkan rating lebih dari 50 kali mempunyai rata rata yang sudah di berikan review tidak lebih dari 24 kali review oleh pengguna.

In [ ]:
# Rekomendasi
- Bekerja samalah dengan publisher Pinguin Books
- Untuk stock buku dengan pengaran JK Rowling dan JRR Tolkien lebih baik di perbanyak
- Stock yang memiliki buku yang sudah di berikan review 50 kali lebih baik di per